decompiler/pipeline/dataflowanalysis/common_subexpression_elimination.py

```diff
@@ -265,18 +261,31 @@ def _is_invalid_dominator(self, basic_block: BasicBlock, expression: Expression)
         usages_in_the_same_block = [usage for usage in self._usages[expression] if usage.block == basic_block]
         return any([isinstance(usage.instruction, Phi) for usage in usages_in_the_same_block])
 
-    def _insert_definition(self, definition: CfgInstruction):
+    def _insert_definition(self, instruction: Instruction, block: BasicBlock, index: int):
         """Insert a new intermediate definition for the given expression at the given location."""
-        definition.block.instructions.insert(definition.index, definition.instruction)
-        for subexpression in _subexpression_dfs(definition.instruction):
-            self._usages[subexpression][definition] += 1
+        block.instructions.insert(index, instruction)
+
+        def update_location(location: CfgInstructionLocation) -> CfgInstructionLocation:
+            if location.block == block and location.index < index:
+                return dataclasses.replace(location, index=location.index + 1)
+            else:
+                return location
+
+        # update positions of expression usages
+        for occurrences in self._usages.values():
+            for location in list(occurrences):
+                if location.block == block and location.index >= index:
+                    occurrences[dataclasses.replace(location, index=location.index + 1)] = occurrences.pop(location)
```

Comment:
Perhaps, this should be a function? And it is a bit hacky, but seems to work.

Reply:
Another idea, but I am not sure whether this is faster/better.  Currently, you iterate over all subexpressions of the cfg, but what if you only iterate over all subexpressions of the basic block? I am not sure whether this is an improvement, probably for long functions, for short I am not sure.

Something similar to the following, probably not exactly. You still have to check whether the location is the same block, so it is perhaps not better. But I have the feeling, we iterate over too many subexpressions and locations here.
```suggestion
        for subexpression in block.subexpressions:
            occurrences = self._usages[subexpression]
            for location in list(occurrences):
                if location.block == block and location.index >= index:
                    occurrences[dataclasses.replace(location, index=location.index + 1)] = occurrences.pop(location)
```

Reply:
Your right that this code is prone to errors and probably doesn't scale nice. After coming back to this I decided to just switch back to my previous solution but instead use an identity check to get the indices of instructions.
This also means that the pull request is now less changes again.